### PR TITLE
[NEB-44] fix: fixed result merge for extended analysis

### PR DESCRIPTION
**Pull request by shinin-light**
[NEB-44]
- [x] Ready for review
- [x] Linked to Jira ticket

#### What does this PR do?

Added the result merge for extended analyses by parsing the sarif results.
All input (i.e. `extendAnalysisOptions.files`) and output (i.e. sarif positions) are always to be considered relative to `baseDir` now for consistency.

[NEB-44]: https://snyksec.atlassian.net/browse/NEB-44

### CODE REVIEW

File: src/analysis.ts [hunk 1]
```diff
@@ -127,36 +127,73 @@ export async function analyzeFolders(options: FileAnalysisOptions): Promise<File
...
+  newAnalysisResults: AnalysisResult,
+  limitToFiles: string[],
+  removedFiles: string[] = [],
+): AnalysisResult {
+  // Start from the new analysis results
+  // For each finding of the old analysis,
+  //  if it's location is not part of the limitToFiles or removedFiles (removedFiles should also be checked against condeFlow),
+  //   append the finding to the new analysis and check if the rule must be added as well
+  const changedFiles = [...limitToFiles, ...removedFiles];
+  const sarifResults = (newAnalysisResults.sarif.runs[0].results || []).filter(res => {
```
Comment (neonnoon) [inline src/analysis.ts:141]: What's the meaning of `newAnalysisResults.sarif.runs[0]`? Specifically reading only the first element of an array kind of raises a flag..

File: src/analysis.ts [hunk 2]
```diff
@@ -127,36 +127,73 @@ export async function analyzeFolders(options: FileAnalysisOptions): Promise<File
...
+      // TODO: This should not be necessary in theory but, in case of two identical files,
+      // Bundle Server returns the finding in both files even if limitToFiles only reports one
+      const loc = res.locations && res.locations[0].physicalLocation?.artifactLocation?.uri;
```
Comment (neonnoon) [inline src/analysis.ts:144]: Just fyi, you can also use optional chaining with arrays `res.locations?.[0]`.

File: src/analysis.ts [hunk 3]
```diff
@@ -127,36 +127,73 @@ export async function analyzeFolders(options: FileAnalysisOptions): Promise<File
...
+      return loc && changedFiles.includes(loc);
+  });
+  const sarifRules = newAnalysisResults.sarif.runs[0].tool.driver.rules || [];
+  const oldResults = oldAnalysisResults.sarif.runs[0].results || [];
+  for (let res of oldResults) {
+    const locations: string[] = (res.locations || []).reduce<string[]>(
```
Comment (neonnoon) [inline src/analysis.ts:150]: Question: Aren't these `reduce`s basically `map().filter()` / `flatMap().filter()`?
No need to change, but might be easier to read.

File: src/files.ts [hunk 4]
```diff
@@ -295,8 +295,10 @@ export async function prepareExtendingBundle(
   let processingFiles: string[] = filterSupportedFiles(files, supportedFiles);
 
   // Exclude files to be ignored based on ignore rules. We assume here, that ignore rules have not been changed.
-  processingFiles = processingFiles.filter(f => !isMatch(f, fileIgnores));
-
+  processingFiles = processingFiles
+    .map(f => resolveBundleFilePath(baseDir, f))
+    .filter(f => !isMatch(resolveBundleFilePath(baseDir, f), fileIgnores));
```
Comment (neonnoon) [inline src/files.ts:300]: This seems weird, the map above maps `processingFiles` with `resolveBundleFilePath`. Why do you need to pass it here again.

Put differently, what is `resolveBundleFilePath(baseDir, resolveBundleFilePath(baseDir, f))`?

**Review by neonnoon — APPROVED** [review]
Approving, as most comments are cosmetic or minor, but I'd appreciate if you could have a look at them.

File: src/analysis.ts (hunk 1 above)
Comment (shinin-light) [reply src/analysis.ts:141]: We only report 1 analysis result per SARIF (the format allows you to report multiple analyses at once)

File: src/analysis.ts (hunk 3 above)
Comment (shinin-light) [reply src/analysis.ts:150]: Great suggestion, It didn't occur to me to simplify it that way

File: src/analysis.ts (hunk 2 above)
Comment (shinin-light) [reply src/analysis.ts:144]: Didn't know, updating

File: src/files.ts (hunk 4 above)
Comment (shinin-light) [reply src/files.ts:300]: It should be idempotent (from testing it), but removed for readability.

File: src/analysis.ts (hunk 3 above)
Comment (shinin-light) [reply src/analysis.ts:161]: Bad news, we need to be compatible with Node 10 and flatMap has been introduced in Node 11... I'm reverting the changes.

**Review by Spoor2709 — APPROVED** [review]
Looks Good to me if you want to comment above what the reduces are doing, I found it semi confusing at first look 😃

**Pull request merged**